test/Validate.SSA.cpp

```diff
@@ -0,0 +1,699 @@
+  ValidateInstructions(str, SPV_ERROR_INVALID_ID);
+}
+
+TEST_F(Validate, ForwardEnqueueKernelNonDominantParameter6Bad) {
+  string str = forwardKernelNonDominantParameterBaseCode + R"(
+                  %err     = OpEnqueueKernel %uintt %dqueue %ndval %nevent %event %revent2 %kfunc %firstp %psize %palign %lsize
+                  %revent2 = OpUndef %eventt
+                             OpReturn
+                             OpFunctionEnd
+                  )";
+  ValidateInstructions(str, SPV_ERROR_INVALID_ID);
+}
+
+TEST_F(Validate, ForwardEnqueueKernelNonDominantParameter8Bad) {
+  string str = forwardKernelNonDominantParameterBaseCode + R"(
+                  %err     = OpEnqueueKernel %uintt %dqueue %ndval %nevent %event %revent %kfunc %firstp2 %psize %palign %lsize
+                  %firstp2 = OpConstant %int8t 0
+                             OpReturn
+                             OpFunctionEnd
+                  )";
+  ValidateInstructions(str, SPV_ERROR_INVALID_ID);
+}
+
+TEST_F(Validate, ForwardEnqueueKernelNonDominantParameter9Bad) {
+  string str = forwardKernelNonDominantParameterBaseCode + R"(
+                  %err    = OpEnqueueKernel %uintt %dqueue %ndval %nevent %event %revent %kfunc %firstp %psize2 %palign %lsize
+                  %psize2 = OpConstant %intt 0
+                            OpReturn
+                            OpFunctionEnd
+                  )";
+  ValidateInstructions(str, SPV_ERROR_INVALID_ID);
+}
+
+TEST_F(Validate, ForwardEnqueueKernelNonDominantParameter10Bad) {
+  string str = forwardKernelNonDominantParameterBaseCode + R"(
+                  %err     = OpEnqueueKernel %uintt %dqueue %ndval %nevent %event %revent %kfunc %firstp %psize %palign2 %lsize
+                  %palign2 = OpConstant %intt 32
+                            OpReturn
+                            OpFunctionEnd
+                  )";
+  ValidateInstructions(str, SPV_ERROR_INVALID_ID);
+}
+
+TEST_F(Validate, ForwardEnqueueKernelNonDominantParameter11Bad) {
+  string str = forwardKernelNonDominantParameterBaseCode + R"(
+                  %err     = OpEnqueueKernel %uintt %dqueue %ndval %nevent %event %revent %kfunc %firstp %psize %palign %lsize2
+                  %lsize2  = OpConstant %intt 1
+                             OpReturn
+                             OpFunctionEnd
+                  )";
+  ValidateInstructions(str, SPV_ERROR_INVALID_ID);
+}
+
+static const bool kWithNDrange = true;
+static const bool kNoNDrange = false;
+pair<string, bool> cases[] = {
+    {"OpGetKernelNDrangeSubGroupCount", kWithNDrange},
+    {"OpGetKernelNDrangeMaxSubGroupSize", kWithNDrange},
+    {"OpGetKernelWorkGroupSize", kNoNDrange},
+    {"OpGetKernelPreferredWorkGroupSizeMultiple", kNoNDrange}};
+
+INSTANTIATE_TEST_CASE_P(KernelArgs, Validate, ::testing::ValuesIn(cases));
+
+string SetOps(string &str, pair<string, bool> param) {
+  regex placeholder("GET_KERNEL_OP");
+  string out = regex_replace(str, placeholder, param.first);
```

Comment:
I think you can accomplish what you want with std::stringstream and its << operator.
And to avoid repeating yourself, you can put the conditional logic in a function that takes a std::ostream* parameter to modify.